zeebe/atomix/cluster/src/main/java/io/atomix/cluster/discovery/DynamicDiscoveryProvider.java

```diff
@@ -0,0 +1,275 @@
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.atomix.cluster.discovery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.Sets;
+import com.google.common.net.HostAndPort;
+import io.atomix.cluster.BootstrapService;
+import io.atomix.cluster.Node;
+import io.atomix.cluster.NodeConfig;
+import io.atomix.cluster.NodeId;
+import io.atomix.utils.event.AbstractListenerManager;
+import io.atomix.utils.net.Address;
+import io.camunda.zeebe.util.VisibleForTesting;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Dynamic cluster node discovery provider.
+ *
+ * <p>This provider resolves DNS addresses to discover nodes in the cluster. It supports:
+ *
+ * <ul>
+ *   <li>Host names that resolve to multiple IP addresses
+ *   <li>IPv4 addresses
+ *   <li>Addresses with or without explicit ports (defaults to MessagingConfig port)
+ *   <li>Periodic DNS resolution refresh to detect changes
+ * </ul>
+ *
+ * <p>The provider periodically refreshes DNS resolutions and fires {@link
+ * NodeDiscoveryEvent.Type#JOIN} events when new nodes are discovered and {@link
+ * NodeDiscoveryEvent.Type#LEAVE} events when nodes are no longer resolved.
+ */
+public final class DynamicDiscoveryProvider
+    extends AbstractListenerManager<NodeDiscoveryEvent, NodeDiscoveryEventListener>
+    implements NodeDiscoveryProvider {
+
+  public static final Type TYPE = new Type();
+  private static final Logger LOGGER = LoggerFactory.getLogger(DynamicDiscoveryProvider.class);
+
+  private final DynamicDiscoveryConfig config;
+  private final Set<Node> discoveredNodes = ConcurrentHashMap.newKeySet();
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private final ScheduledExecutorService scheduler =
+      Executors.newSingleThreadScheduledExecutor(
+          runnable -> {
+            final Thread thread = new Thread(runnable, "dynamic-discovery");
+            thread.setDaemon(true);
+            return thread;
+          });
+
+  private final int defaultPort;
+  private ScheduledFuture<?> refreshTask;
+  private final Function<String, List<InetAddress>> addressResolver;
```

Comment:
```suggestion
  private final Function<String, List<InetAddress>> hostnameResolver;
```

:wrench: addressResolver actually resolves hostname, not addresses, so it's name it's a bit misleading. Considering that the input type is `String` it's not clear that it should not resolve and address (i.e. host + port).